### DISCUSSION TITLE
How do I integrate NextJS' OpenTelemetry support with .Net Aspire?

### DISCUSSION THREAD

**Discussion by KennethHoff** [microsoft/aspire#5304] [Q&A] [answered]
I've tried implementing support for OpenTelemetry w/ NextJS using .Net Aspire. I spent a fair number of hours yesterday looking through NextJS' docs (https://nextjs.org/docs/app/building-your-application/optimizing/open-telemetry) as well as yours (https://github.com/dotnet/aspire-samples/tree/b013f64e063dce2a4c72b840eab83d43991a5d22/samples/AspireWithNode), but no matter what I do, I am unable to see anything pop up in the `traces` panel.

There are so many moving pieces that it's near-impossible to figure out what is going from; All the environment variables are there, and according to NextJS' docs, this is all you should have to do to follow the standard systems:

```ts
import { registerOTel } from '@vercel/otel';

// The following function does get called by NextJS, according to the console.
export function register() {
  console.log('service name:', process.env.OTEL_SERVICE_NAME); // This is the right string
  console.log('endpoint:', process.env.OTEL_EXPORTER_OTLP_ENDPOINT); // This seems to be the right string.
  registerOTel(); // This is supposed to use these env. variables, according to NextJS' docs.
  console.log('Registered OTel'); // This gets called.
}

```

This is my Aspire code:

```c#
var builder = DistributedApplication.CreateBuilder(args);
builder
	.AddNpmApp(
		"Frontend",
		"<PATH>",
		"dev"
	)
	.WithOtlpExporter()
	.WithHttpEndpoint(env: "PORT", port: 3000)
	.WithExternalHttpEndpoints();

var app = builder.Build();

await app.RunAsync().ConfigureAwait(false);

```

As you can see in this image, the traces selector doesn't even include the service.

<img width="1693" alt="Screenshot 2024-08-15 at 09 18 28" src="https://github.com/user-attachments/assets/38738327-8acb-42e5-aa96-1407fa5a8042">

**Comment by davidfowl** [comment]
You might need to make the default OTLP endpoint http based to make these scenarios work more seamlessly.

cc @JamesNK @samsp-msft @DamianEdwards

**Reply by JamesNK** [reply]
We can have both. The only thing required is including an env var in launchSettings.json with the randomly generated OTLP HTTP port.

**Reply by KennethHoff** [reply]
What does all of this mean for my current scenario? Can I change this myself somehow, or does this require a change to Aspire itself?

**Reply by davidfowl** [reply]
In your launchSettings for the apphost project, there should be an endpoint specified as an environment variable DOTNET_DASHBOARD_OTLP_ENDPOINT_URL, change that to be http instead of https.

**Reply by JamesNK** [reply]
You can do it. If you're using 8.2, then you should just need to add this line of config for your launchSettings.json for the app host:

https://github.com/dotnet/aspire/blob/9a59d122949de428321599c503da59c54fb673d6/playground/BrowserTelemetry/BrowserTelemetry.AppHost/Properties/launchSettings.json#L12

I need to do some work to document this better.

**Reply by KennethHoff** [reply]
Just for completeness, is that PORT randomly chosen by you now, randomly chosen by Aspire, or is it always the same one?

**Reply by JamesNK** [reply]
It's randomly generated when the file is created. You can use the values in the launchSettings.json I linked to, or pick your own. It doesn't matter.

**Reply by KennethHoff** [reply]
I've yet to see whether this is a fix, but assuming it does, how come it didn't exist in the template? Was it added in 8.2? (I made mine in 8.1)

**Reply by KennethHoff** [reply]
I added this line now and removed the other line\*, and it still doesn't seem to work. The environment variables look correct; the protocol is 'http/protobuf', and endpoint is correct, but nothing arrives.


\* I originally didn't remove the gRPC endpoint, but that seemingly has higher priority, so it didn't use the new one, and there's no option to choose which protocol to use.

**Reply by enemyss** [reply]
Hi, I'm experiencing same exact issue with Aspire v8.2 and next js v14.2.9, @KennethHoff  did you manage to configure it properly?

**Reply by KennethHoff** [reply]
Nope. I've given up for now 😭

**Reply by JamesNK** [reply]
https://github.com/dotnet/aspire/tree/9a59d122949de428321599c503da59c54fb673d6/playground/BrowserTelemetry is an example of doing this.

If you can't get it to work and can create a simple repo, someone on the team can take a look.

**Reply by JamesNK** [reply]
I double checked that receiving browser telemetry still works. Demo:

![browser-telemetry-8 2](https://github.com/user-attachments/assets/ec8f4913-69f9-4a65-bbc7-bda46e339ad2)

**Reply by KennethHoff** [reply]
NextJS is not Browser, it's a NodeJS application with their own telemetry system that (supposedly) interacts with OTel

https://nextjs.org/docs/app/building-your-application/optimizing/open-telemetry

**Reply by JamesNK** [reply]
Ok. I don't know of anything special about NodeJS that stops it working. If anything it should be more likely to work because browsers must also worry about CORS while NodeJS doesn't.

**Reply by davidfowl** [reply]
@KennethHoff did you try changing the default OTLP endpoint to be http instead of https? Does the nextjs otel exporter support grpc? Those are likely the 2 things you need to do to make it work.

**Reply by DamianEdwards** [reply]
There's a [sample of a Node.js app sending telemetry in the samples repo](https://github.com/dotnet/aspire-samples/tree/main/samples/AspireWithNode). The Node app had to be explicitly configured to disable SSL certificate validation as Node.js doesn't trust the ASP.NET Core dev certificate which is protecting the OTLP endpoint: https://github.com/dotnet/aspire-samples/blob/ab410c06fddf23f1e0369c2c6124df3f5f4909be/samples/AspireWithNode/AspireWithNode.AppHost/Program.cs#L20

**Reply by enemyss** [reply]
Hi, I made simple repo with aspire and next.js - https://github.com/enemyss/aspire-next-js

**Reply by enemyss** [reply]
Looks like grpc protocol is not supported.
With `OTEL_LOG_LEVEL=debug` you can see the actual issue
`@vercel/otel: Unsupported OTLP traces protocol: grpc. Using http/protobuf.`
I switched to http/protobuf and I'm seeing traces in aspire's dashboard👍

**Reply by KennethHoff** [reply]
How did you "switch to http/protobuf"?

**Reply by enemyss** [reply]
By removing "DOTNET_DASHBOARD_OTLP_ENDPOINT_URL" from launchSettings.json, so there is only     "DOTNET_DASHBOARD_OTLP_HTTP_ENDPOINT_URL" left

**Reply by KennethHoff** [reply]
I figured it out. The `OTEL_LOG_LEVEL=debug` hint helped a lot.

In addition to requiring the `http/protobuf` protocol, I also had to disable TLS certification, using `.WithEnvironment("NODE_TLS_REJECT_UNAUTHORIZED", "0")`

**Reply by KennethHoff** [reply]
I figured out how to make it work with the gRPC endpoint as well:

```ts
import { AzureMonitorTraceExporter } from '@azure/monitor-opentelemetry-exporter';
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-grpc';
import { BatchSpanProcessor } from '@opentelemetry/sdk-trace-node';
import { registerOTel } from "@vercel/otel";

const spanProcessors = [new BatchSpanProcessor(new OTLPTraceExporter())];

if (process.env.APPLICATIONINSIGHTS_CONNECTION_STRING) {
  spanProcessors.push(new BatchSpanProcessor(new AzureMonitorTraceExporter()));
}

registerOTel({ spanProcessors: spanProcessors });

```

**Reply by enemyss** [reply]
thanks, I managed to use grpc protocol as well, reused your code but had issue using @vercel/otel registration, but it works well with manual one

```
import { OTLPTraceExporter } from "@opentelemetry/exporter-trace-otlp-grpc";
import { Resource } from '@opentelemetry/resources'
import { BatchSpanProcessor } from "@opentelemetry/sdk-trace-node";
import { NodeSDK } from '@opentelemetry/sdk-node'
import { ATTR_SERVICE_NAME } from '@opentelemetry/semantic-conventions'
 
const sdk = new NodeSDK({
  resource: new Resource({
    [ATTR_SERVICE_NAME]: process.env.OTEL_SERVICE_NAME,
  }),
  spanProcessor: new BatchSpanProcessor(new OTLPTraceExporter()),
})
sdk.start()
```

**Reply by KennethHoff** [reply, marked as answer]
This is my actual code:

```ts
// instrumentation.ts


import { registerOTel } from '@vercel/otel';

export async function register() {
  if (process.env.NEXT_RUNTIME === 'nodejs') {
    const { spanProcessors } = await import('./instrumentation.node');
    registerOTel({ spanProcessors: spanProcessors });
  }
}

// instrumentation.node.js


import { AzureMonitorTraceExporter } from '@azure/monitor-opentelemetry-exporter';
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-grpc';
import { BatchSpanProcessor } from '@opentelemetry/sdk-trace-node';

const spanProcessors = [new BatchSpanProcessor(new OTLPTraceExporter())];

if (process.env.APPLICATIONINSIGHTS_CONNECTION_STRING) {
  spanProcessors.push(new BatchSpanProcessor(new AzureMonitorTraceExporter()));
}

export { spanProcessors };

```

**Reply by oising** [reply]
Hey @KennethHoff  -- how did you get it working for HTTP? I tried replacing exporter-trace-otlp-grpc with exporter-trace-otlp-http but it started blowing up with 415 unsupported media type from the endpoint, and it wasn't really clear what was happening. I got GRPC working fine on the NodeJS side but if you want to include client-side OTEL using otel/js, you have to switch _everything_ to http, including the backend (limitation of Aspire)

**Reply by KennethHoff** [reply]
I can't fully remember what I ended up with, but it stopped working at some point, and I haven't looked into it

**Reply by timfewi** [reply]
@oising I had the same issue and managed to solve it by making the following change:

```diff
- import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-http';
+ import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-proto';
```

```typescript

// instrumentation.node.ts 
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-proto';
import { Resource } from '@opentelemetry/resources';
import { NodeSDK } from '@opentelemetry/sdk-node';
import { SimpleSpanProcessor } from '@opentelemetry/sdk-trace-node';
import { ATTR_SERVICE_NAME } from '@opentelemetry/semantic-conventions';

function parseOtlpHeaders(headerString: string): Record<string, string> {
    return headerString.split(',')
        .map(pair => pair.trim())
        .reduce((headers, pair) => {
            const [key, value] = pair.split('=');
            if (key && value) {
                headers[key.trim()] = value.trim();
            }
            return headers;
        }, {} as Record<string, string>);
}

const otlpEndpoint = `${process.env.OTEL_EXPORTER_OTLP_ENDPOINT}/v1/traces`;

const additionalHeaders = process.env.OTEL_EXPORTER_OTLP_HEADERS
    ? parseOtlpHeaders(process.env.OTEL_EXPORTER_OTLP_HEADERS)
    : {};

const exporterHeaders = {
    'Content-Type': 'application/x-protobuf',
    ...additionalHeaders
};

const exporter = new OTLPTraceExporter({
    url: otlpEndpoint,
    headers: exporterHeaders
});

const sdk = new NodeSDK({
    resource: new Resource({
        [ATTR_SERVICE_NAME]: process.env.OTEL_SERVICE_NAME || 'frontend'
    }),
    spanProcessors: [new SimpleSpanProcessor(exporter)]
});

sdk.start();

```

```typescript
// instrumentation.ts
export async function register() {
    if (process.env.NEXT_RUNTIME === 'nodejs') {
        await import('./instrumentation.node');
    }
}

```

**Comment by ByronMayne** [comment]
I have the same issue was @oising, when using the javascript exporter you get a ` 415 (Unsupported Media Type)` using the json version. I have a TypeScript frontend application and I want to push the telemetry directly to the dashboard.  This is my setup

```ts
import { BatchSpanProcessor, WebTracerProvider } from '@opentelemetry/sdk-trace-web';
import { DocumentLoadInstrumentation } from '@opentelemetry/instrumentation-document-load';
import { ZoneContextManager } from '@opentelemetry/context-zone';
import { registerInstrumentations } from '@opentelemetry/instrumentation';
import { XMLHttpRequestInstrumentation } from '@opentelemetry/instrumentation-xml-http-request/build/src/xhr';
import { OTLPTraceExporter } from "@opentelemetry/exporter-trace-otlp-http";
import { UserInteractionInstrumentation } from '@opentelemetry/instrumentation-user-interaction';
import { ATTR_SERVICE_NAME } from "@opentelemetry/semantic-conventions";
import { resourceFromAttributes, type Resource } from '@opentelemetry/resources';

const endpoint = import.meta.env.OTEL_EXPORTER_OTLP_ENDPOINT;
const serviceName = import.meta.env.OTEL_SERVICE_NAME;

const attributes = {
    [ATTR_SERVICE_NAME]: serviceName,
    ...parseDelimitedValues(import.meta.env.OTEL_RESOURCE_ATTRIBUTES)
}

const headers = {
    "Content-Type":  'application/json',
    ...parseDelimitedValues(import.meta.env.OTEL_EXPORTER_OTLP_HEADERS)
};



const exporter = new OTLPTraceExporter({
    url: `${endpoint}/v1/traces`,
    headers: headers,
    concurrencyLimit: 10,
    timeoutMillis: 30000,
});


const provider = new WebTracerProvider({
    resource: resourceFromAttributes(attributes),
    spanProcessors: [
        new BatchSpanProcessor(exporter, {
            maxQueueSize: 100,
            maxExportBatchSize: 10,
            scheduledDelayMillis: 500,
            exportTimeoutMillis: 30000,
        })]
});


provider.register({
  // Changing default contextManager to use ZoneContextManager - supports asynchronous operations - optional
    contextManager: new ZoneContextManager(),
});

registerInstrumentations({
    instrumentations: [
        new UserInteractionInstrumentation(),
        new DocumentLoadInstrumentation(),
        new XMLHttpRequestInstrumentation()
    ],
});

function parseDelimitedValues(s:string) {
    const headers:string[] = s.split(','); // Split by comma
    const o:Record<string, string> = {};

    headers.forEach(header => {
        const [key, value] = header.split('='); // Split by equal sign
        o[key.trim()] = value.trim(); // Add to the object, trimming spaces
    });

    return o;
}

```

**Reply by ByronMayne** [reply]
As a heads up for anyone else wasting their time on this. It turns out the 415 mime error is because they never implemented JSON. https://github.com/dotnet/aspire/blob/3bbae260225f7dabbc2c09e0f5315d5e48644718/src/Aspire.Dashboard/Otlp/Http/OtlpHttpEndpointsBuilder.cs#L160. This goes against what all the docs say.

**Reply by davidfowl** [reply]
cc @JamesNK @IEvangelist

**Reply by oising** [reply]
Yes, this is correct. I ended up using gRPC.

**Reply by JamesNK** [reply]
Where do the docs say JSON OTLP is supported? I can fix that.

Aspire dashboard has only ever supported gRPC and HTTP/Protobuf.

**Reply by ByronMayne** [reply]
I think it's just a confusion from the docs on my part. There is the grpc endpoint and the http endpoint I falsely assumed that http is json and proto is for grpc. There are exporters for both JSON and proto for tracing, logging, and metrics. 

Maybe just add an explicit call-out and an error message along with a return code. For example saying only 'application/x-proto' is an accepted content time. The only reason I figured it out was browsing the source code and attaching the debugger.  I also tripped up because the code is setup to parse the application/json content type but at the very bottom just ignores it. 

TLDR: the issue was my fault but there is room for improvement to make sure no one else falls into the trap.

**Reply by JamesNK** [reply]
Ok. AI is on it: https://github.com/dotnet/aspire/issues/11253

**Reply by ByronMayne** [reply]
Thank you :)
